include/mdw_dslash5_tensor_core.cuh

```diff
@@ -0,0 +1,475 @@
+    b = __hmul2(sm_b[(threadIdx.y * 4 + 3) * N_sm_d2 + 3 * threadIdx.x + 0], max_i_div_max2_);
+    c = __hmul2(sm_b[(threadIdx.y * 4 + 3) * N_sm_d2 + 3 * threadIdx.x + 1], max_i_div_max2_);
+    d = __hmul2(sm_b[(threadIdx.y * 4 + 3) * N_sm_d2 + 3 * threadIdx.x + 2], max_i_div_max2_);
+    vector_store(&out[sid + 2 * output.volumeCB], 0, __4half22integer8_rn<storage_vec>(a, b, c, d));
+#else
+
+    typedef typename VectorType<storage_type, 4>::type storage_vec;
+    storage_vec *out = reinterpret_cast<storage_vec *>(output.field);
+    half2 a, b;
+
+    a = __hmul2(sm_b[(threadIdx.y * 4 + 0) * N_sm_d2 + 3 * threadIdx.x + 0], max_i_div_max2_);
+    b = __hmul2(sm_b[(threadIdx.y * 4 + 0) * N_sm_d2 + 3 * threadIdx.x + 1], max_i_div_max2_);
+    out[sid + 0 * output.volumeCB] = __2half22integer4_rn<storage_vec>(a, b);
+
+    a = __hmul2(sm_b[(threadIdx.y * 4 + 0) * N_sm_d2 + 3 * threadIdx.x + 2], max_i_div_max2_);
+    b = __hmul2(sm_b[(threadIdx.y * 4 + 1) * N_sm_d2 + 3 * threadIdx.x + 0], max_i_div_max2_);
+    out[sid + 1 * output.volumeCB] = __2half22integer4_rn<storage_vec>(a, b);
+
+    a = __hmul2(sm_b[(threadIdx.y * 4 + 1) * N_sm_d2 + 3 * threadIdx.x + 1], max_i_div_max2_);
+    b = __hmul2(sm_b[(threadIdx.y * 4 + 1) * N_sm_d2 + 3 * threadIdx.x + 2], max_i_div_max2_);
+    out[sid + 2 * output.volumeCB] = __2half22integer4_rn<storage_vec>(a, b);
+
+    a = __hmul2(sm_b[(threadIdx.y * 4 + 2) * N_sm_d2 + 3 * threadIdx.x + 0], max_i_div_max2_);
+    b = __hmul2(sm_b[(threadIdx.y * 4 + 2) * N_sm_d2 + 3 * threadIdx.x + 1], max_i_div_max2_);
+    out[sid + 3 * output.volumeCB] = __2half22integer4_rn<storage_vec>(a, b);
+
+    a = __hmul2(sm_b[(threadIdx.y * 4 + 2) * N_sm_d2 + 3 * threadIdx.x + 2], max_i_div_max2_);
+    b = __hmul2(sm_b[(threadIdx.y * 4 + 3) * N_sm_d2 + 3 * threadIdx.x + 0], max_i_div_max2_);
+    out[sid + 4 * output.volumeCB] = __2half22integer4_rn<storage_vec>(a, b);
+
+    a = __hmul2(sm_b[(threadIdx.y * 4 + 3) * N_sm_d2 + 3 * threadIdx.x + 1], max_i_div_max2_);
+    b = __hmul2(sm_b[(threadIdx.y * 4 + 3) * N_sm_d2 + 3 * threadIdx.x + 2], max_i_div_max2_);
+    out[sid + 5 * output.volumeCB] = __2half22integer4_rn<storage_vec>(a, b);
+
+#endif
+  }
+
+  // For "reload" version(reload == true) of wmma gemm, matrix a is loaded when
+  // needed.
+  // It is a waste of time but has less register usage.
+  // For "preload" version(reload == false) of wmma gemm, matrix a is preloaded
+  // before hand.
+  // It saves time but uses more registers.
+  template <int BlockDimX, int Ls, int M, int N, int M_sm, int N_sm, bool reload, class T>
+  __device__ inline void wmma_gemm(T *a_frag, half *sm_a, half *sm_b, half *sm_c)
+  {
+    constexpr int WMMA_M = 16;
+    constexpr int WMMA_N = 16;
+    constexpr int WMMA_K = 16;
+
+    constexpr int tm_dim = M / WMMA_M;
+    constexpr int tn_dim = N / WMMA_N;
+
+    constexpr int total_warp = BlockDimX * Ls / 32;
```

Comment:
Probably an overkill comment, but we may want to replace this hard coded `32` with some sort of `warp_size` variable.